server/src/main/java/io/crate/expression/scalar/ArraySliceFunction.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.crate.expression.scalar.array.ArrayArgumentValidators.ensureInnerTypeIsNotUndefined;
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public class ArraySliceFunction extends Scalar<List<Object>, Object> {
+
+    public static final String NAME = "array_slice";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature(),
+                parseTypeSignature("array(E)")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            ArraySliceFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private ArraySliceFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        ensureInnerTypeIsNotUndefined(boundSignature.getArgumentDataTypes(), signature.getName().name());
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public List<Object> evaluate(TransactionContext txnCtx, NodeContext nodeContext, Input<Object>[] args) {
+        List<Object> listInput = (List<Object>) args[0].value();
+        if (listInput == null || listInput.isEmpty()) {
+            return List.of();
+        }
+
+        int from = args.length >= 2 && args[1].value() != null ?
+            ((Number) args[1].value()).intValue() - 1 :
+            0;
+        if (from >= listInput.size()) {
+            return List.of();
+        }
+
+        int to = args.length == 3 && args[2].value() != null ?
+            ((Number) args[2].value()).intValue() :
+            listInput.size();
```

Comment:
Please call `.value()` only once for each argument. The input could be anything, e.g. a function, where `.value()` may cause expensive evaluations.

Reply:
https://github.com/crate/crate/pull/11949/commits/74d2ed1e691196f264d8233e355ae6b772a0a3bf#diff-df78ac731cf947e6fd6c503f5180febab00006d5e3955f0ce3cc3f8c78b28c01R81-R101